sphinx/config.rst
.. $Id$
  
..
  This file is part of pyFormex 1.0.2  (Thu Jun 18 15:35:31 CEST 2015)
  pyFormex is a tool for generating, manipulating and transforming 3D
  geometrical models by sequences of mathematical operations.
  Home page: http://pyformex.org
  Project page:  http://savannah.nongnu.org/projects/pyformex/
  Copyright 2004-2015 (C) Benedict Verhegghe (benedict.verhegghe@feops.com)
  Distributed under the GNU General Public License version 3 or later.
  
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see http://www.gnu.org/licenses/.
  
  
.. include:: defines.inc
.. include:: links.inc

.. _cha:config:

********************
Configuring pyFormex
********************

Many aspects of pyFormex can be configured to better suit the user's needs and likings.
These can range from merely cosmetic changes to important extensions of the
functionality. As is written in a scripting language and distributed as source,
the user can change every single aspect of the program. And the GNU-GPL license
under which the program is distributed guarantees that you have access to the
source and are allowed to change it.

Most users however will only want to change minor aspects of the program, and
would rather not have to delve into the source to do just that. Therefore we
have gathered some items of that users might like to change, into separate files
where thay can easily be found. Some of these items can even be set
interactivley through the GUI menus.

Often users want to keep their settings between subsequent invocation of the
program. To this end, the user preferences have to be stored on file when
leaving the program and read back when starting the next time. While it might
make sense to distinct between the user's current settings in the program and
his default preferences, the current configuration system of (still under
development) does not allow such distinction yet. Still, since the topic is so
important to the user and the configuration system in is already quite complex,
we tought it was necessary to provide already some information on how to
configure. Be aware though that important changes to this system will likely
occur.


.. _sec:pyf-conf-files:

Configuration files
===================

On startup, reads its configurable data from a number of files. Often there are
not less than four configuration files, read in sequence. The settings in each
file being read override the value read before. The different configuration
files used serve different purposes. On a typical GNU/Linux installation, the
following files will be read in sequence:

* PYFORMEX-INSTALL-PATH/pyformexrc: this file should never be changed , neither
  by the user nor the administrator. It is there to guarantee that all settings
  get an adequate default value to allow to correctly start up.

* /etc/pyformex: this file can be used by the system administrator to make
  system-wide changes to the installation. This could e.g. be used to give all
  users at a site access to a common set of scripts or extensions.

* /.pyformexrc: this is where the user normally stores his own default
  settings.

* CURRENT-DIR/.pyformex: if the current working directory from which is started
  contains a file named .pyformex, it will be read too. This makes it possible to
  keep different configurations in different directories, depending on the
  purpose. Thus, one directory might aim at the use of for operating on
  triangulated surfaces, while another might be intended for pre- and post-
  processing of Finite Element models.

* Finally, the ``--config=`` command line option provides a way to specify
  another file with any name to be used as the last configuration file.

On exit,will store the changed settings on the last user configuration file that
was read. The first two files mentioned above are system configuration files and
will never be changed by the program. A user configuration file will be
generated if none existed.

.. warning:: Currently, when pyFormex exits, it will just dump all
  the changed configuration (key,value) pairs on the last
  configuration file, together with the values it read from that
  file. pyFormex will not detect if any changes were made to that
  file between reading it and writing back. Therefore, the user should
  never edit the configuration files directly while pyFormex is still
  running. Always close the program first!


.. _sec:syntax-conf-files:

Syntax of the configuration files
=================================

All configuration files are plain text files where each non blank line is one of
the following:

* a comment line, starting with a '#',

* a section header, of the form '[section-name]',

* a valid Python instruction.

The configuration file is organized in sections. All lines preceding the first
section name refer to the general (unnamed) section.

Any valid Python source line can be used. This allows for quite complex
configuration instructions, even importing Python modules. Any line that binds a
value to a variable will cause a corresponding configuration variable to be set.
The user can edit the configuration files with any text editor, but should make
sure the lines are legal Python. Any line can use the previously defined
variables, even those defined in previously read files.

In the configuration files, the variable pyformexdir refers to the directory
where was installed (and which is also reported by the ``pyformex --whereami``
command).


.. _sec:conf-vars:

Configuration variables
=======================

Many configuration variables can be set interactively from the GUI, and the user
may prefer to do it that way. Some variables however can not (yet) be set from
th GUI. And real programmers may prefer to do it with an editor anyway. So here
are some guidelines for setting some interesting variables. The user may take a
look at the installed default configuration file for more examples.

General section
"""""""""""""""

* ``syspath = []``: Value is a list of path names that will be appended to
  the Python's sys.path variable on startup. This enables your scripts
  to import modules from other than default Python paths.

* ``scriptdirs = [ ('Examples',examplesdir),
  ('MyScripts',myscriptsdir)]``: a list of tuples (name,path). On
  startup, all these paths will be scanned for scripts and these will
  be added in the menu under an item named name.

.. index:: single: autorun

* ``autorun = '.pyformex.startup'``: name of a script that will be
  executed on startup, before any other script (specified on the
  command line or started from the GUI).

* ``editor = 'kedit'``: sets the name of the editor that will be used
  for editing pyformex scripts.

* ``viewer = 'firefox'``: sets the name of the html viewer to be used to
  display the html help screens.

* ``browser = 'firefox'``: sets the name of the browser to be used to access the
  website.

* ``uselib = False``: do not use the acceleration library. The default (True) is to use it when it is available.


Section ``[gui]``
"""""""""""""""""

.. index:: single: splash image

* ``splash = 'path-to-splash-image.png'``: full path name of the image to be used
  as splash image on startup.

* ``modebar = True``: adds a toolbar with the render mode buttons. Besides
  True or False, the value can also be one of 'top', 'bottom', 'left'
  or 'right', specifying the placement of the render mode toolbar at
  the specified window border. Any other value that evaluates True
  will make the buttons get included in the top toolbar.

* ``viewbar = True``: adds a toolbar with different view
  buttons. Possioble values as explained above for modebar.

* ``timeoutbutton = True``: include the timeout button in the toolbar. The
  timeout button, when depressed, will cause input widgets to time out
  after a prespecified delay time. This feature is still experimental.

* ``plugins = ['surface_menu', 'formex_menu', 'tools_menu']``: a list of
  plugins to load on startup. This is mainly used to load extra
  (non-default) menus in the GUI to provide extended
  functionality. The named plugins should be available in the
  'plugins' subdirectory of the installation. To autoload user
  extensions from a different place, the autorun script can be used.

.. End



History
# $Id:$
##
##  This file is part of the pyFormex project.
##  pyFormex is a tool for generating, manipulating and transforming 3D
##  geometrical models by sequences of mathematical operations.
##  Home page: http://pyformex.org
##  Project page:  https://savannah.nongnu.org/projects/pyformex/
##  Copyright (C) Benedict Verhegghe (benedict.verhegghe@ugent.be) 
##  Distributed under the GNU General Public License version 3 or later.
##
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see http://www.gnu.org/licenses/.
##

2009-06 BV: Released pyFormex 0.8

2008-04 BV: Released pyFormex 0.7

2008-02 BV: Started the pyFormex User Meeting

2007-11 BV: Released pyFormex 0.6

2007-08 BV: Released pyFormex 0.5, a major step towards a multifunctional
            program.

2007-02 BV: Released pyFormex 0.4.2, including lots of work on surface
            models.

2006-11-03 BV: Released pyFormex 0.4. This release is marked by the transition
               to the Qt4 widget set.

2006-04-15 BV: Released pyFormex 0.3

2006: pyFormex was used in some of our research projects and proved to be
      a valuable tool. This necessitated and inspired new development work
      on pyFormex. Lots of new features were implemented, too many to be
      listed here.
      Some of the milestones:
      - pyFormex was used in the design of wire stents (see WireStent example).
        The process involved the automatic creation of Finite Element
        Analysis jobs and automatic processing of the results.
      - pyFormex was used to measure geometries obtained from medical images
        and to create a patient specific geometrical hardware model. This
        initiated our further work with surface models.
      - pyFormex was used by civil engineering students for their projects
        in designing shell structures. Some of these will be included as
        examples in future releases. 


2005-12-09 BV : First anniversary release 0.2.2
2005-04-23 BV : Interim 0.2.1 release with lots of new stuff
2005-01-03 BV : First public release on berlios.de (0.2)
2004-12-09 BV : Registered project pyFormex on berlios.de
2004-07-08 BV : Created a first release (0.1.1)
2004-07-02 BV : Created a Subversion repository
2004-06-29 BV :	After reading about Formian, I quickly hacked together
                the Formex class

# End

pyformex/config.py
# $Id$
##
##  This file is part of pyFormex 1.0.2  (Thu Jun 18 15:35:31 CEST 2015)
##  pyFormex is a tool for generating, manipulating and transforming 3D
##  geometrical models by sequences of mathematical operations.
##  Home page: http://pyformex.org
##  Project page:  http://savannah.nongnu.org/projects/pyformex/
##  Copyright 2004-2015 (C) Benedict Verhegghe (benedict.verhegghe@feops.com)
##  Distributed under the GNU General Public License version 3 or later.
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see http://www.gnu.org/licenses/.
##
"""A general yet simple configuration class.

| (C) 2005 Benedict Verhegghe
| Distributed under the GNU GPL version 3 or later

Why
   I wrote this simple class because I wanted to use Python
   expressions in my configuration files. This is so much more fun
   than using .INI style config files.  While there are some other
   Python config modules available on the web, I couldn't find one
   that suited my needs and my taste: either they are intended for
   more complex configuration needs than mine, or they do not work
   with the simple Python syntax I expected.

What
   Our Config class is just a normal Python dictionary which can hold
   anything.  Fields can be accessed either as dictionary lookup
   (config['foo']) or as object attributes (config.foo).  The class
   provides a function for reading the dictionary from a flat text
   (multiline string or file). I will always use the word 'file'
   hereafter, because that is what you usually will read the
   configuration from.  Your configuration file can have named
   sections. Sections are stored as other Python dicts inside the top
   Config dictionary. The current version is limited to one level of
   sectioning.
"""
from __future__ import print_function

from pyformex.mydict import Dict
from pyformex.utils import formatDict
import copy


class Config(Dict):
    """A configuration class allowing Python expressions in the input.

    The configuration settings are stored in the __dict__ of a Python
    object.  An item 'foo' in the configuration 'config' can be accessed
    either as dictionary lookup (``config['foo']``) or as object attribute
    (``config.foo``).

    The configuration object can be initialized from a multiline string or
    a text file (or any other object that allows iterating over strings).

    The format of the config file/text is described hereafter.

    All config lines should have the format: key = value, where key is a
    string and value is a Python expression The first '=' character on the
    line is the delimiter between key and value.  Blanks around both the
    key and the value are stripped.  The value is then evaluated as a
    Python expression and stored in a variable with name specified by the
    key. This variable is available for use in subsequent configuration
    lines. It is an error to use a variable before it is defined.  The
    key,value pair is also stored in the config dictionary, unless the key
    starts with an underscore ('_'): this provides for local variables.

    Lines starting with '#' are comments and are ignored, as are empty
    and blank lines.  Lines ending with '\' are continued on the next
    line.  A line starting with '[' starts a new section. A section is
    nothing more than a Python dictionary inside the config
    dictionary. The section name is delimited by '['and ']'. All
    subsequent lines will be stored in the section dictionary instead
    of the toplevel dictionary.

    All other lines are executed as python statements. This allows
    e.g. for importing modules.

    Whole dictionaries can be inserted at once in the config with the
    update() function.

    All defined variables while reading config files remain available
    for use in the config file statements, even over multiple calls to
    the read() function. Variables inserted with addSection() will not
    be available as individual variables though, but can be access as
    ``self['name']``.

    As an example, if your config file looks like::

       aa = 'bb'
       bb = aa
       [cc]
       aa = 'aa'
       _n = 3
       rng = range(_n)

    the resulting configuration dictionary is
    ``{'aa': 'bb', 'bb': 'bb', 'cc': {'aa': 'aa', 'rng': [0, 1, 2]}}``

    As far as the resulting Config contents is concerned, the following are
    equivalent::

       C.update({'key':'value'})
       C.read("key='value'\\n")

    There is an important difference though: the second line will make a
    variable key (with value 'value') available in subsequent Config read()
    method calls.
    """

    def __init__(self,data={},default=None):
        """Creates a new Config instance.

        The configuration can be initialized with a dictionary, or
        with a variable that can be passed to the read() function.
        The latter includes the name of a config file, or a multiline string
        holding the contents of a configuration file.
        """
        Dict.__init__(self, default=default)
        if isinstance(data, dict):
            self.update(data)
        elif data:
            self.read(data)


    def update(self,data={},name=None,removeLocals=False):
        """Add a dictionary to the Config object.

        The data, if specified, should be a valid Python dict.
        If no name is specified, the data are added to the top dictionary
        and will become attributes.
        If a name is specified, the data are added to the named attribute,
        which should be a dictionary. If the name does not specify a
        dictionary, an empty one is created, deleting the existing attribute.

        If a name is specified, but no data, the effect is to add a new
        empty dictionary (section) with that name.

        If removeLocals is set, keys starting with '_' are removed from the
        data before updating the dictionary and not
        included in the config. This behaviour can be changed by setting
        removeLocals to false.
        """
        if removeLocals:
            for k in list(data.keys()):
                if k[0] == '_':
                    del data[k]
        if name:
            if name not in self or not isinstance(self[name], dict):
                self[name] = Dict()
            self[name].update(data)
        else:
            Dict.update(self, data)


    def _read_error(self, filename, lineno, line):
        if filename:
            where = 'config file %s,' % filename
        else:
            where = ''
        raise RuntimeError('Error in %s line %d:\n%s' % (where, lineno, line))


    def read(self,fil,debug=False):
        """Read a configuration from a file or text

        `fil` is a sequence of strings. Any type that allows a loop like
        ``for line in fil:``
        to iterate over its text lines will do. This could be a file type, or
        a multiline text after splitting on '\\n'.

        The function will try to react intelligently if a string is passed as
        argument. If the string contains at least one '\\n', it will be
        interpreted as a multiline string and be splitted on '\\n'.
        Else, the string will be considered and a file with that name will
        be opened. It is an error if the file does not exist or can not be
        opened.

        The function returns self, so that you can write: cfg = Config().
        """
        filename = None
        if isinstance(fil, str):
            if fil.find('\n') >= 0:
                fil = fil.split('\n')
            else:
                filename = fil
                fil = open(fil, 'r')
        section = None
        contents = {}
        lineno = 0
        continuation = False
        comments = False
        for line in fil:
            lineno += 1
            ls = line.strip()
            if comments:
                comments = ls[-3:] != '"""'
                ls = ''
            else:
                comments = ls[:3] == '"""'

            if comments or len(ls)==0 or ls[0] == '#':
                continue

            if continuation:
                s += ls
            else:
                s = ls

            continuation = s[-1] == '\\'
            if s[-1] == '\\':
                s = s[:-1]

            if continuation:
                continue

            if s[0] == '[':
                if contents:
                    self.update(name=section, data=contents, removeLocals=True)
                    contents = {}
                i = s.find(']')
                if i<0:
                    self.read_error(filename, lineno, line)
                section = s[1:i]
                if debug:
                    print("Starting new section '%s'" % section)
                continue
            else:
                if debug:
                    print("READ: "+line)
            i = s.find('=')
            if i >= 0:
                key = s[:i].strip()
                if len(key) == 0:
                    self.read_error(filename, lineno, line)
                contents[key] = eval(s[i+1:].strip())
                globals().update(contents)
            else:
                exec(s)
        if contents:
            self.update(name=section, data=contents, removeLocals=True)
        return self


    def __setitem__(self, key, val):
        """Allows items to be set as self[section/key] = val.

        """
        i = key.rfind('/')
        if i == -1:
            self.update({key:val})
        else:
            self.update({key[i+1:]:val}, key[:i])


    def __getitem__(self, key):
        """Allows items to be addressed as self[key].

        This is equivalent to the Dict lookup, except that items in
        subsections can also be retrieved with a single key of the format
        section/key.
        While this lookup mechanism works for nested subsections, the syntax
        for config files allows for only one level of sections!
        Also beware that because of this functions, no '/' should be used
        inside normal keys and sections names.
        """
        i = key.rfind('/')
        if i == -1:
            return Dict.__getitem__(self, key)
        else:
            try:
                return self[key[:i]][key[i+1:]]
            except KeyError:
                return self._default_(key)


    def __delitem__(self, key):
        """Allows items to be delete with del self[section/key].

        """
        i = key.rfind('/')
        if i == -1:
            Dict.__delitem__(self, key)
        else:
            del self[key[:i]][key[i+1:]]


    def __str__(self):
        """Format the Config in a way that can be read back.

        This function is mostly used to format the data for writing it to
        a configuration file. See the write() method.

        The return value is a multiline string with Python statements that can
        be read back through Python to recreate the Config data. Usually
        this is done with the Config.read() method.
        """
        s = ''
        for k, v in self.items():
            if not isinstance(v, Dict):
                s += formatDict({k:v})
        for k, v in self.items():
            if isinstance(v, Dict):
                s += "\n[%s]\n" % k
                s += formatDict(v)
        return s


    def write(self,filename,header="# Config written by pyFormex    -*- PYTHON -*-\n\n",trailer="\n# End of config\n"):
        """Write the config to the given file

        The configuration data will be written to the file with the given name
        in a text format that is both readable by humans and by the
        Config.read() method.

        The header and trailer arguments are strings that will be added at
        the start and end of the outputfile. Make sure they are valid
        Python statements (or comments) and that they contain the needed
        line separators, if you want to be able to read it back.
        """
        fil = open(filename, 'w')
        fil.write(header)
        fil.write("%s" % self)
        fil.write(trailer)
        fil.close()


    def keys(self,descend=True):
        """Return the keys in the config.

        By default this descends one level of Dicts.
        """
        keys = list(Dict.keys(self))
        if descend:
            for k, v in self.items():
                if isinstance(v, Dict):
                    keys += ['%s/%s' % (k, ki) for ki in v]

        return keys


if __name__ == '__main__':


    def show(s):
        try:
            v = eval(s)
            print("%s = %s" % (s, v))
        except:
            print("%s ! ERROR" % s)

    C = Config("""# A simple config example
aa = 'bb'
bb = aa
[cc]
aa = 'aa'    # yes ! comments are allowed (they are stripped by eval())
_n = 3       # local: will get stripped
rng = range(_n)
""")
    show("C")
    show("C['aa']")
    show("C['cc']")
    show("C['cc/aa']")
    show("C['dd']")


    def reflookup(key):
        return C[key]

    D = Config(default = reflookup)

    show("D")
    show("D['aa']")
    show("D['cc']")
    show("D['cc/aa']")
    show("D['dd']")

    D['aa'] = 'wel'
    D['dd'] = 'hoe'
    D['cc/aa'] = 'ziedewel'
    show("D")
    show("C")
    show("D['cc/aa']")
    show("D['cc/rng']")
    print("BUT!!!!")
    show("D['cc']")

    # This should give an error
    show("D['ee']")
    show("D.get('ee','NO Error')")
    show("D.get('cc/ee','NO Error')")

    D['cc/bb'] = 'ok'
    show("D.keys()")
    del D['aa']
    del D['cc/aa']
    show("D.keys()")
    del D['cc']
    show("D.keys()")


# End
